tephra-core/src/test/java/co/cask/tephra/persist/LocalTransactionStateStorageTest.java

```diff
@@ -49,4 +66,99 @@ protected Configuration getConfiguration(String testName) throws IOException {
+  // TransactionEditV2 is not TransactionEdit. Since the code path this test is verifying is the same path between
+  // HDFS and Local Storage, having this only over here is fine.
+  @SuppressWarnings("deprecation")
+  @Test
+  public void testLongTxnBackwardsCompatibility() throws Exception {
+    Configuration conf = getConfiguration("testLongTxnBackwardsCompatibility");
+
+    // Use SnapshotCodec version 1
+    String latestSnapshotCodec = conf.get(TxConstants.Persist.CFG_TX_SNAPHOT_CODEC_CLASSES);
+    conf.set(TxConstants.Persist.CFG_TX_SNAPHOT_CODEC_CLASSES, DefaultSnapshotCodec.class.getName());
+
+    // Create v2 TransactionEdit
+    class TransactionEditV2 extends TransactionEdit {
+      public TransactionEditV2(long writePointer, long visibilityUpperBound, State state, long expirationDate, 
+                               Set<ChangeId> changes, long commitPointer, boolean canCommit, TransactionType type) {
+        super(writePointer, visibilityUpperBound, state, expirationDate, changes, commitPointer, canCommit, type);
+      }
+      @Override
+      public void write(DataOutput out) throws IOException {
+        new TransactionEditCodecV2().encode(this, out);
+      }
+    };
+
+    TransactionStateStorage storage = null;
+    try {
+      storage = getStorage(conf);
+      storage.startAndWait();
+
+      // Create transaction snapshot and transaction edits with version when long running txns had -1 expiration.
+      Collection<Long> invalid = Lists.newArrayList();
+      NavigableMap<Long, TransactionManager.InProgressTx> inProgress = Maps.newTreeMap();
+      long time1 = System.currentTimeMillis();
+      inProgress.put(time1, new TransactionManager.InProgressTx(time1 - 5, -1));
+      long time2 = time1 + 100;
+      inProgress.put(time2, new TransactionManager.InProgressTx(time2 - 50, time2 + 1000));
+      Map<Long, Set<ChangeId>> committing = Maps.newHashMap();
+      Map<Long, Set<ChangeId>> committed = Maps.newHashMap();
+      TransactionSnapshot snapshot = new TransactionSnapshot(time1, 0, time1, invalid,
+                                                             inProgress, committing, committed);
+      long time3 = time1 + 200;
+      TransactionEdit edit1 = new TransactionEditV2(time3, time3 - 10, TransactionEdit.State.INPROGRESS, -1,
+                                                    null, 0L, false, null);
+      long time4 = time1 + 300;
+      TransactionEdit edit2 = new TransactionEditV2(time4, time4 - 10, TransactionEdit.State.INPROGRESS, time4 + 1000,
+                                                    null, 0L, false, null);
+
+      // write snapshot and transaction edit
+      storage.writeSnapshot(snapshot);
+      TransactionLog log = storage.createLog(time1);
+      log.append(edit1);
+      log.append(edit2);
+      log.close();
+
+      // Start transaction manager
+      conf.setInt(TxConstants.Manager.CFG_TX_CLEANUP_INTERVAL, 3);
+      conf.setInt(TxConstants.Manager.CFG_TX_TIMEOUT, 2);
```

Comment:
I don't think we should adjust the timeout values and just leave out the timeout portion of the test.

This can be done better after we port over TimestampOracle, which allows you to deterministically control the observed time.